# numpy array의 생김새

In [1]:
import numpy

In [2]:
array1 = numpy.array([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31])

In [3]:
array1

array([ 2,  3,  5,  7, 11, 13, 17, 19, 23, 29, 31])

In [4]:
type(array1)

numpy.ndarray

In [5]:
array1.shape

(11,)

In [7]:
array2 = numpy.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])

In [8]:
type(array2)

numpy.ndarray

In [9]:
array2.shape

(3, 4)

In [10]:
array1.size

11

In [11]:
array2.size

12

# numpy array를 만드는 다양한 방법

## 1. 파이썬 리스트를 통해 생성
: numpy 모듈의 array 메소드에 파라미터로 파이썬 리스트를 넘겨주면 numpy array가 리턴됩니다.

In [12]:
array1 = numpy.array([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31])
print(array1)

[ 2  3  5  7 11 13 17 19 23 29 31]


## 2. 균일한 값으로 생성

### numpy 모듈의 full 메소드를 사용하면, 모든 값이 같은 numpy array를 생성할 수 있습니다.

In [13]:
array1 = numpy.full(6, 7)
print(array1)

[7 7 7 7 7 7]


### 모든 값이 0인 numpy array 생성
: 모든 값이 0인 numpy array를 생성하기 위해서는 full 메소드를 사용하면 되겠죠. 하지만 사실은 더 간편한 방법이 있습니다.

In [15]:
array1 = numpy.full(6, 0)
array2 = numpy.zeros(6, dtype=int)

print(array1)
print()
print(array2)

[0 0 0 0 0 0]

[0 0 0 0 0 0]


### 모든 값이 1인 numpy array 생성
: 모든 값이 1인 numpy array를 생성하는 것도 비슷합니다. zeros 메소드 대신 ones를 사용하면 됩니다.

In [16]:
array1 = numpy.full(6, 1)
array2 = numpy.ones(6, dtype=int)

print(array1)
print()
print(array2)

[1 1 1 1 1 1]

[1 1 1 1 1 1]


## 3. 랜덤한 값들로 생성
: 어쩔 때는 임의의 값들로 배열을 생성시키고 싶습니다. 그럴 때는 numpy의 random 모듈의 random 함수를 사용하면 됩니다.

numpy 모듈 안에 random이라는 모듈이 있고, 그 안에 또 random이라는 함수가 있는 겁니다.

In [17]:
array1 = numpy.random.random(6)
array2 = numpy.random.random(6)

print(array1)
print()
print(array2)

[0.45018194 0.08611388 0.00634141 0.61787589 0.30158644 0.75849759]

[0.77389303 0.83095232 0.46363343 0.27334812 0.61772734 0.33356263]


## 4. 연속된 값들이 담긴 numpy array 생성
: numpy 모듈의 arange 함수를 사용하면 연속된 값들이 담겨 있는 numpy array를 생성할 수 있습니다.

arange 함수는 파이썬의 기본 함수인 range와 굉장히 비슷한 원리로 동작하는데요. 파라미터가 1개인 경우, 2개인 경우, 3개인 경우 모두 살펴봅시다.

### 파라미터 1개
: arange(m)을 하면 0부터 m-1까지의 값이 담긴 numpy array가 리턴됩니다.

In [18]:
array1 = numpy.arange(6)
print(array1)

[0 1 2 3 4 5]


### 파라미터 2개
: arange(n, m)을 하면 n부터 m-1까지의 값들이 담긴 numpy array가 리턴됩니다.

In [19]:
array1 = numpy.arange(2, 7)
print(array1)

[2 3 4 5 6]


### 파라미터 3개
: arange(n, m, s)를 하면 n부터 m-1까지의 값들 중 간격이 s인 값들이 담긴 numpy array가 리턴됩니다.

In [20]:
array1 = numpy.arange(3, 17, 3)
print(array1)

[ 3  6  9 12 15]


# 모듈 별명 지어주기

In [21]:
import numpy as np

array1 = np.full(6, 0)
array2 = np.zeros(6)
array3 = np.full(6, 6)
array4 = np.ones(9)
array5 = np.random.random(6)
array6 = np.random.random(10)

In [22]:
import matplotlib.pyplot as plt

# 인덱싱, 슬라이싱

In [24]:
import numpy as np

In [26]:
array1 = np.array([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31])

In [27]:
array1[0]

2

In [28]:
array1[2]

5

In [29]:
array1[-1]

31

In [30]:
array1[-2]

29

In [31]:
array1[[1, 3, 4]]

array([ 3,  7, 11])

In [32]:
array2 = np.array([2, 1, 3])

In [33]:
array1[array2]

array([5, 3, 7])

In [35]:
array1[2:7]

array([ 5,  7, 11, 13, 17])

In [36]:
array1[:7]

array([ 2,  3,  5,  7, 11, 13, 17])

In [37]:
array1[2:]

array([ 5,  7, 11, 13, 17, 19, 23, 29, 31])

In [38]:
array1[2:11]

array([ 5,  7, 11, 13, 17, 19, 23, 29, 31])

In [40]:
array1[2:11:3]

array([ 5, 13, 23])

# numpy 기본 연산

In [41]:
import numpy as np

In [42]:
array1 = np.arange(10)
array2 = np.arange(10, 20)

In [43]:
array1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [44]:
array2

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

In [45]:
array1 * 2

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [46]:
array1 / 2

array([0. , 0.5, 1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5])

In [48]:
array1 + 2

array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [49]:
array1 ** 2

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81], dtype=int32)

In [50]:
array1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [51]:
array1 = array1 * 2

In [52]:
array1

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [53]:
array1 = np.arange(10)
array2 = np.arange(10, 20)

In [54]:
array1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [55]:
array2

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

In [56]:
array1 + array2

array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28])

In [57]:
array1 * array2

array([  0,  11,  24,  39,  56,  75,  96, 119, 144, 171])

In [58]:
array1 / array2

array([0.        , 0.09090909, 0.16666667, 0.23076923, 0.28571429,
       0.33333333, 0.375     , 0.41176471, 0.44444444, 0.47368421])

# numpy 불린 연산

In [59]:
import numpy as np

In [60]:
array1 = np.array([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31])

In [61]:
array1 > 4

array([False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [62]:
array1 % 2 == 0

array([ True, False, False, False, False, False, False, False, False,
       False, False])

In [64]:
booleans = np.array([True, True, False, True, True, False, True, True, True, False, True])

In [65]:
np.where(booleans)

(array([ 0,  1,  3,  4,  6,  7,  8, 10], dtype=int64),)

In [66]:
array1 > 4

array([False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [67]:
np.where(array1 > 4)

(array([ 2,  3,  4,  5,  6,  7,  8,  9, 10], dtype=int64),)

In [68]:
filter = np.where(array1 > 4)
filter

(array([ 2,  3,  4,  5,  6,  7,  8,  9, 10], dtype=int64),)

In [70]:
array1[filter]

array([ 5,  7, 11, 13, 17, 19, 23, 29, 31])

# numpy 기본 통계

## 최댓값, 최솟값
: max 메소드와 min 메소드를 사용하면 numpy array의 최댓값과 최솟값을 구할 수 있습니다.

In [72]:
import numpy as np

array1 = np.array([14, 6, 13, 21, 23, 31, 9, 5])

print(array1.max())
print(array1.min())

31
5


## 평균값
: mean 메소드를 사용하면 numpy array의 평균값을 구할 수 있습니다.

In [73]:
import numpy as np

array1 = np.array([14, 6, 13, 21, 23, 31, 9, 5])
print(array1.mean())

15.25


## 중앙값
: median 메소드를 사용하면 중간값을 구할 수 있는데요. 특이하게 median은 numpy array의 메소드가 아니라 numpy의 메소드입니다.

In [74]:
import numpy as np

array1 = np.array([8, 12, 9, 15, 16])
array2 = np.array([14, 6, 13, 21, 23, 31, 9, 5])

print(np.median(array1))
print(np.median(array2))

12.0
13.5


 - array1을 정렬하면 중앙값이 12입니다.
 - array2에는 짝수개의 요소가 있기 때문에 중앙값이 13, 14 두 개입니다. 둘의 평균값을 내면 13.5 입니다.

### 표준 편차, 분산
: 표준 편차와 분산은 값들이 평균에서 얼마나 떨어져 있는지 나타내는 지표입니다. 

In [76]:
import numpy as np
array1 = np.array([14, 6, 13, 21, 23, 31, 9, 5])

print(array1.std())
print(array1.var())

8.496322733983215
72.1875
